curl: enable HTTP/3 support by default

NyCodeGHG opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
See https://curl.se/docs/http3.html

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

NyCodeGHG committed:
curl: enable HTTP/3 support by default

NyCodeGHG committed:
{ngtcp2, curl, nghttp2}: use openssl instead of quictls

NyCodeGHG committed:
nghttp3: replace fetchFromGitHub with fetchurl

Otherwise we run into infrec.

NyCodeGHG committed:
curlHTTP3: drop

Izorkin commented:
If I'm not mistaken, OpenSSL's implementation of the QUIC protocol has performance issues compared to QuicTLS.

Izorkin commented:
> But isn't this using ngtcp2 for QUIC and only OpenSSL for crypto?
> Or am I misunderstanding some of the build options?

The ngtcp2 package supports OpenSSL 3.5 only through the `libngtcp2_crypto_ossl` helper library. I don't think this is full support.
I would prefer to use quictls in curl by default, but no one will agree to that - https://github.com/NixOS/nixpkgs/pull/179579

emilazy reviewed (commented):
Seems like a good idea, as discussed on Matrix.

I agree that ngtcp2 + OpenSSL 3.5 seems like the obvious route here. AIUI, it does not have the issues of OpenSSL’s native QUIC implementation.

Scrumplex reviewed (approved):
Changes LGTM. Thanks for making sure that every commit still evaluates.

Izorkin commented:
This PR release changes the QUIC protocol implementation in ngtcp2 from OpenSSL 3.5, which is not supported by other projects. This may lead to bugs.

NyCodeGHG commented:
ngtcp2 + OpenSSL seems to be still marked as experimental, not sure if we actually want to ship this then
https://github.com/ngtcp2/ngtcp2/blob/2334b653df8a8ec9bafe6a1f2bf3ed30db6e9ef1/README.rst?plain=1#L69

alois31 commented:
Samba already builds against GnuTLS, so it would probably use `ngtcp2-gnutls`, in which case it doesn't matter at all what TLS library the main `ngtcp2` package uses. I do not know about the kernel module, but would be very surprised if it depended at all on the ngtcp2 userspace library. Indeed I should have been clearer in the previous comment about the experimental status, however it is not so clear to me what to make of that given that curl (the only user of the `ngtcp2` package) documents building ngtcp2 against OpenSSL without this disclaimer.

emilazy commented:
> If necessary, quictls can be replaced with wolfSSL, BoringSSL, aws-lc or libressl.

From reading the curl `configure.ac`, It seems like these would require switching the main TLS library used by curl; it cannot use ngtcp2 + AWS‐LC for HTTP/3 while using OpenSSL for everything else, say.

So this would be a question of changing the default TLS library used by curl in general. GnuTLS we would rather get rid of as much as possible, LibreSSL lags behind OpenSSL these days, BoringSSL comes with no stability promises and is disclaimed by its upstream as unsuitable for general‐purpose use. The only palatable options would seem to be wolfSSL or AWS‐LC. AWS‐LC is in the OpenSSL fork family and already moderately load‐bearing for Rust stuff, so likely to be the most viable choice here.

Still, though, I am sceptical we would want to make that switch when we use OpenSSL so extensively anyway. curl and nghttp2 are the only consumers of our ngtcp2 package and curl seems to prefer the ngtcp2 + OpenSSL combination. It’s also the only option they’re giving [CodeQL coverage in CI](https://github.com/curl/curl/pull/18557), for instance. I think the choice here is clear if we want to enable HTTP/3 support by default, which I think we do.

(After writing this, I see that it appears the CMake build of curl can support multiple TLS libraries, but [it is explicitly not supported with HTTP/3](https://github.com/curl/curl/blob/e2d3b832445a05ed2e590fa5a40b0914f932bc42/docs/INSTALL.md#multissl-and-http3). Even if it was, I’m not sure we’d want two of them in the closure with the attendant additional security exposure, and we don’t use the CMake build of curl anyway.)

NyCodeGHG committed:
ngtcp2: 1.14.0 -> 1.15.1, use fetchurl

fetchurl instead of fetchFromGitHub, because otherwise we run into infrec

Diff: https://github.com/ngtcp2/ngtcp2/compare/v1.14.0...v1.15.1

Changelog: https://github.com/ngtcp2/ngtcp2/releases/tag/v1.15.1

NyCodeGHG committed:
nghttp2: 1.65.0 -> 1.67.1

Diff: https://github.com/nghttp2/nghttp2/compare/v1.65.0...v1.67.1

Changelog: https://github.com/nghttp2/nghttp2/releases/tag/v1.67.1

emilazy commented on pkgs/by-name/ng/nghttp2/package.nix:
```diff
@@ -68,15 +67,14 @@ stdenv.mkDerivation rec {
       c-aresMinimal
       libev
       zlib
+      openssl
     ]
-    ++ lib.optionals (enableApp && !enableHttp3) [ openssl ]
```
Is it correct to not gate on `enableApp` any more?

Izorkin commented:
Maybe then should create a separate package `ngtcp2-openssl` and use it?

emilazy reviewed (approved):
Confirmed that `curl --http3-only https://google.com/` works on `aarch64-darwin`.

NyCodeGHG commented:
Maybe relevant question from curl maintainer Daniel Stenberg: https://curl.se/mail/lib-2025-10/0000.html
